### ISSUE TITLE
peer certificate cannot be authenticated 

### ISSUE THREAD

**Issue by Glurak** [megous/megatools#329]
```
C:\Program Files (x86)\megatools>megadl "***"

ERROR: Download failed for '***': API call 'g' failed: HTTP POST failed: CURL error: Peer certificate cannot be authenticated with given CA certificates

C:\Program Files (x86)\megatools>megadl --version
megatools 1.9.99git - command line tools for Mega.nz

Written by Ondřej Jirman <megous@megous.com>, 2013
Go to http://megatools.megous.com for more information
```

**Comment by megous** [comment]
Megatools use system certificate store. So either you're missing CA certificate that issued server cert for mega.nz there, or there's active MITM attack somewhere between your computer and mega.nz. That could be anything from antivirus, to proxy trying to decrypt your data.

**Comment by Robby-** [comment]
I'm also having this issue. I shouldn't be missing any CA certs as I'm applying all updates provided by my operating system vendor. Pretty sure there is also no MITM taking place.

Any way to make Megatools use a `ca-certificates.crt` file as it did in version 1.9.96? Or an option to ignore invalid certificates, similar to wget's `--no-check-certificate`?

**Comment by megous** [comment]
I think it's better to identify and solve the underlying issue.

Download curl and do:

    curl -v https://eu.api.mega.co.nz/cs

For me it outputs:

```
* TCP_NODELAY set
* Connected to eu.api.mega.co.nz (31.216.147.134) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=NZ; postalCode=1010; ST=Auckland; L=Auckland; street=15, Pwc Tower, 188 Quay Street,; O=Mega Limited; OU=PremiumSSL Wildcard; CN=*.api.mega.co.nz
*  start date: Dec 11 00:00:00 2017 GMT
*  expire date: Dec 10 23:59:59 2020 GMT
*  subjectAltName: host "eu.api.mega.co.nz" matched cert's "*.api.mega.co.nz"
*  issuer: C=GB; ST=Greater Manchester; L=Salford; O=COMODO CA Limited; CN=COMODO RSA Organization Validation Secure Server CA
*  SSL certificate verify ok.
```

**Comment by Robby-** [comment]
I'm using Windows 7 SP1 64-bit. I have verified that the Windows certificate store has the `COMODO RSA Certification Authority` root certificate, and verified it is also the same serial (`‎4c aa f9 ca db 63 6f e0 1f f7 4e d8 5b 03 86 9d`) as in the chain the eu.api.mega.co.nz API server uses.

For me, using a Windows version of curl, this outputs:
```
*   Trying 2001:67c:1998:202::184...
* TCP_NODELAY set
* Connected to eu.api.mega.co.nz (2001:67c:1998:202::184) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: N:\Programs\CLI\curl\curl-ca-bundle.crt
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=NZ; postalCode=1010; ST=Auckland; L=Auckland; street=15, Pwc Tower, 188 Quay Street,; O=Mega Limited; OU=PremiumSSL Wildcard; CN=*.api.mega.co.nz
*  start date: Dec 11 00:00:00 2017 GMT
*  expire date: Dec 10 23:59:59 2020 GMT
*  subjectAltName: host "eu.api.mega.co.nz" matched cert's "*.api.mega.co.nz"
*  issuer: C=GB; ST=Greater Manchester; L=Salford; O=COMODO CA Limited; CN=COMODO RSA Organization Validation Secure Server CA
*  SSL certificate verify ok.
```

Obviously curl uses its own certificate store, so this curl verification succeeds.

**Comment by Robby-** [comment]
Hmm, I just noticed by the way, that my error message is different than OP's message:
`ERROR: Download failed for 'https://mega.nz/<...>': API call 'g' failed: HTTP POST failed: CURL error: Peer certificate cannot be authenticated with given CA certificates`
But maybe the issues are related nonetheless.

**Comment by Robby-** [comment]
Is Megatools's included curl version actually using the Windows certificate store?

**Comment by megous** [comment]
I compile libcurl for megatools with:

```sh
build() {
  cd "$srcdir/curl-$pkgver"

  for _arch in ${_architectures}; do
    rm -rf build-${_arch}
    mkdir build-${_arch}
    pushd build-${_arch}
    mingw_configure \
      --with-winssl \
      --with-zlib=/usr/${_arch} \
      --with-libidn=/usr/${_arch} \
      --disable-ldap \
      --disable-file \
      --disable-ftp \
      --disable-dict \
      --disable-gopher \
      --disable-imap \
      --disable-imaps \
      --disable-pop3 \
      --disable-pop3s \
      --disable-rtsp \
      --disable-smb \
      --disable-smbs \
      --disable-smtp \
      --disable-smtps \
      --disable-telnet \
      --disable-tftp
    make
    popd
  done

}
```

**Comment by megous** [comment]
Interesting that curl for windows is not itself using Windows TLS implementation/certificate store.

**Comment by megous** [comment]
BTW, that causes "issues" with some anti-* programs, because they install snooping MITM proxies and custom certificates for those proxies.

It's not really an issue, because that's exactly what this whole encryption/certificate stuff is designed to prevent, so I don't mind going back, except that that will make anti-* program users come out of woodwork again, telling me I should fix that by using window certificate store.

No win for me. :D

**Comment by Robby-** [comment]
If it helps, I'm getting the curl binaries from here:
* https://bintray.com/vszakats/generic/curl/ 
* https://github.com/vszakats/curl-for-win

**Comment by megous** [comment]
I'd build CURL_VERBOSE enabled megatools for windows for debugging, but cross-compiler in arch linux is currently broken. I'll try it later.

**Comment by Robby-** [comment]
According to [this](https://curl.haxx.se/docs/sslcerts.html) page, `curl -V` output should contain `WinSSL` in the output.
Quote from the above page:
`If the version string says "WinSSL" in it, then it was built with Schannel support.`

This is my `curl.exe -V` output:
```
curl 7.58.0 (x86_64-pc-win32) libcurl/7.58.0 OpenSSL/1.1.0g (WinSSL) zlib/1.2.11 WinIDN libssh2/1.8.0 nghttp2/1.29.0
Release-Date: 2018-01-24
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile SSPI Kerberos SPNEGO NTLM SSL libz TLS-SRP HTTP2 HTTPS-proxy MultiSSL 
```

Despite WinSSL being there, it also seems to use the .crt file. Or maybe its just logging it but not using it.

From what I see you specify `--with-winssl`, so I do not know whats going wrong. Debugmode might help indeed.

Thanks. :)

**Comment by megous** [comment]
Hmm, the whole windows cross-compiler suite got dropped from the official Arch Linux repos for now. It's too much work to get it working again. I won't be able to do much about this anytime soon.

**Comment by Robby-** [comment]
Oh, that makes me a sad panda.

**Comment by Robby-** [comment]
Maybe you could use one of these binary building services, such as Bintray, as a workaround?

**Comment by Robby-** [comment]
Hm, or not, their service seems to be more of an addition to other services, like [Travis CI](https://travis-ci.org/) and [AppVeyor](https://www.appveyor.com/).

**Comment by megous** [comment]
There's no workaround, I have all deps in PKGBUILDs, tied closely to Arch Linux packaging system. The sanest approach is to build the mingw toolchain packages from AUR.

**Comment by megous** [comment]
I've managed to get the cross-toolchain running from AUR. Try the 1.9.99git build from the website. It has new option --debug=http, which enables CURLOPT_VERBOSE

**Comment by Robby-** [comment]
I didn't try the new option yet, but I first went to try to use `megadl` like before, and to my surprise it worked (since you didn't mention any other changes made). No certificate errors. The new 1.9.99git build works fine.

So, I don't have any useful debug output to show now. However, due to this new debug option I can now confirm `Schannel` is being used, because I see a bunch of `schannel: <...>` messages appear.

Whatever else was changed (aside from adding the new debug option) between the old 1.9.99git build and the new 1.9.99git build seems to have fixed the issue. So, I suppose this issue can be closed, then.

Thank you for (unintentionally? :P) fixing it.
